File: 09/taehyuk/684.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        class OptimizedDisjointSet:
            def __init__(self, n):
                self.parent = list(range(n))
                self.rank = [1] * n

            def find(self, u):
                if u != self.parent[u]:
                    self.parent[u] = self.find(self.parent[u])
                return self.parent[u]

            def merge(self, u, v):
                u = self.find(u)
                v = self.find(v)
                if u != v:
                    if self.rank[u] > self.rank[v]:
                        self.parent[v] = u
                    elif self.rank[u] < self.rank[v]:
                        self.parent[u] = v
                    else:
                        self.parent[v] = u
                        self.rank[u] += 1

        n = len(edges)
        disjointSet = OptimizedDisjointSet(n + 1)

        for u, v in edges:
            if disjointSet.find(u) == disjointSet.find(v):
                return [u, v]
            else:
                disjointSet.merge(u, v)

        return []

File: 09/ahnjaewoo/684.cpp
class Solution {
public:
    unordered_map<int, int> parent;

    int findParent(int node) {
        if (parent.find(node) == parent.end() || parent[node] == node) {
            parent[node] = node;
            return node;
        }
        else return findParent(parent[node]);
    }

    int merge(int node1, int node2) {
        if (node1 == node2) return 1;
        else if (findParent(node1) == findParent(node2)) return 0;
        parent[findParent(node1)] = findParent(node2);
        return 1; 
    }

    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        vector<int> answer;
        for (int i = 0; i < edges.size(); i++) {
            int node1 = edges[i][0];
            int node2 = edges[i][1];
            if (merge(node1, node2) == 0) answer = {node1, node2};
        }
        return answer;
    }
};